plugins/coreboot/fu-plugin-coreboot.c

```diff
@@ -0,0 +1,154 @@
+{
+	fu_plugin_add_rule (plugin, FU_PLUGIN_RULE_SUPPORTS_PROTOCOL, "org.coreboot.fmap");
+
+	/* make sure that flashrom plugin is ready to receive devices */
+	fu_plugin_add_rule (plugin, FU_PLUGIN_RULE_RUN_AFTER, "org.flashrom");
+	fu_plugin_set_build_hash (plugin, FU_BUILD_HASH);
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+}
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	g_autofree const gchar *triplet = NULL;
+	g_autofree gchar *name = NULL;
+	const gchar *major;
+	const gchar *minor;
+	const gchar *vendor;
+	const gchar *version;
+	g_autoptr(FuDevice) dev;
+	GBytes *bios_table;
+	gboolean updatable = TRUE;
+	gint firmware_size = 0;
+	struct fmap *fmap;
+
+	vendor = fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_BIOS_VENDOR);
+	if (!vendor)
+		return FALSE;
+
+	version = fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_BIOS_VERSION);
+	if (version) {
+		triplet = fu_plugin_coreboot_version_string_to_triplet (version, error);
+		/* Ignore developer builds */
+		if (g_strrstr (version, "dirty"))
+			updatable = FALSE;
+	}
+
+	if (!version || !triplet) {
+		major = fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_BIOS_MAJOR_RELEASE);
+		if (!major) {
+			g_set_error (error,
+					G_IO_ERROR,
+					G_IO_ERROR_INVALID_DATA,
+					"Missing BIOS major release");
+			return FALSE;
+		}
+		minor = fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_BIOS_MINOR_RELEASE);
+		if (!minor) {
+			g_set_error (error,
+					G_IO_ERROR,
+					G_IO_ERROR_INVALID_DATA,
+					"Missing BIOS minor release");
+			return FALSE;
+		}
+		triplet = g_strdup_printf ("%s.%s.0", major, minor);
+	}
+
+	if (!triplet)
+		return FALSE;
+	dev = fu_device_new ();
+
+	fu_device_set_version (dev, triplet, FWUPD_VERSION_FORMAT_TRIPLET);
+	fu_device_set_summary (dev, "Open Source system boot firmware");
+	fu_device_set_id (dev, vendor);
+	fu_device_set_vendor (dev, vendor);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_add_icon (dev, "computer");
+        name = fu_plugin_coreboot_get_name_for_type (plugin, NULL);
+	if (!name)
+		name = fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_PRODUCT_NAME);
+	fu_device_set_name (dev, name);
+	fu_device_set_vendor (dev, fu_plugin_get_dmi_value (plugin, FU_HWIDS_KEY_MANUFACTURER));
+	fu_device_add_instance_id (dev, "main-system-firmware");
+
+	fu_device_set_metadata (dev, FU_DEVICE_METADATA_FLASHROM_DEVICE_KIND, "system-firmware");
+
+	bios_table = fu_plugin_get_smbios_data (plugin, FU_SMBIOS_STRUCTURE_TYPE_BIOS);
+	if (bios_table) {
+		guint32 bios_characteristics;
+		gsize len;
+		const guint8 *value = g_bytes_get_data (bios_table, &len);
+		if (len >= 0x9) {
+			firmware_size = (value[0x9] + 1) * 64 * 1024;
+		}
+		if (len >= (0xa + sizeof(guint32))) {
+			memcpy (&bios_characteristics, &value[0xa], sizeof (guint32));
+			if (!(bios_characteristics & (1 << 11))) // "BIOS is upgradeable (Flash)"
+				updatable = FALSE;
+		}
+	}
+
+	if (updatable)
+		fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_UPDATABLE);
```

Comment:
A nice policy that has been adopted in other plugins is that if it's not updatable that an update error is set indicating why (and what you can do to get it updatable if possible).  I think it would be good to to bring that in here.